sling/task/learner.cc

```diff
@@ -0,0 +1,116 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sling/task/learner.h"
+
+#include <unistd.h>
+
+#include "sling/task/task.h"
+
+namespace sling {
+namespace task {
+
+using namespace myelin;
+
+void LearnerTask::Train(Task *task, myelin::Network *model) {
+  // Get training parameters.
+  task->Fetch("epochs", &epochs_);
+  task->Fetch("report_interval", &report_interval_);
+  task->Fetch("rampup", &rampup_);
+
+  // Initialize statistics counters.
+  num_workers_ = task->GetCounter("workers");
+  num_epochs_total_ = task->GetCounter("epochs_total");
+  num_epochs_completed_ = task->GetCounter("epochs_completed");
+
+  epoch_ = 0;
+  num_epochs_total_->Increment(epochs_);
+
+  // Start training threads.
+  int threads = task->Get("workers", jit::CPU::Processors());
+  WorkerPool pool;
+  pool.Start(threads, [this, model](int index) {
+    sleep(index * rampup_);
+    num_workers_->Increment();
+    Worker(index, model);
+  });
+
+  // Evaluate model on regular intervals. The workers signal when it is time
+  // for the next eval round or training has completed.
+  for (;;) {
+    // Wait for next eval or completion.
+    {
+      std::unique_lock<std::mutex> lock(eval_mu_);
+      eval_model_.wait(lock);
+    }
+
+    // Run evaluation.
+    Evaluate(epoch_, model);
+
+    // Check if we are done.
+    if (epoch_ >= epochs_) break;
+  }
+
+  // Wait until workers completes.
```

Comment:
Is 1 epoch == 1 mini-batch?

Say we have 10 worker threads, each processing their own mini-batches. Say the first 5 finish early, and that brings the total number of mini-batches (=epochs) to epochs_. So the main thread would run evaluation, but the other 5 threads are still busy updating the network parameters. So the eval might be slightly off. Further,  the parameters we save at the end are also slightly different from the ones used for evaluation (since other threads might update the params further during pool.Join()). I just want to ensure whether this minor issue is expected, and ignored for the sake of implementation simplicity.

Reply:
The evals are running while the parameters are being updated in parallel, but this is by design. However, I have added a final eval after all the workers complete, so the final eval will always be done on the final model. I have also implemented a 1 sec polling interval for the ramp period to prevent these from stalling completion of training.